### ISSUE TITLE
1.9.3 consumes a lot of RAM if spell-checker dictionaries are installed

### ISSUE THREAD

**Issue by 0136** [telegramdesktop/tdesktop#6983]
### Steps to reproduce
1. Installed Debian 9.
2. Installed Telegram, had been happily using it for some years, everything was fine.
3. Updated Telegram to 1.9.3 (by pressing the "Update" button in Telegram itself, as I usually do).

### Expected behaviour
Telegram consumes a moderate amount of physical memory (under several hundred MB).
Telegram starts quickly (in a few seconds).

### Actual behaviour after the update to 1.9.3

Telegram consumes a lot of physical memory (~2.5 GiB, as displayed by `top` in the `RES` column).
Telegram starts slowly (~1 minute or so).

If I start Firefox in addition to Telegram, the system almost runs out of memory (my laptop has 4 GiB of physical RAM installed), and becomes essentially unusable until I kill the Telegram process.

### Configuration
**Operating system:**
Debian 9, Xfce 4.12

**Version of Telegram Desktop:**
1.9.3

**Used theme**:
Classic

### More details

1.
As I've figured out, my Debian system had dozens of spell-checker dictionaries installed (a dictionary for every supported locale). I guess this is a fairly typical/default setup. (Or is it just me?)

2.
I've turned off "System -> Advanced -> Use system spell checker", and restarted Telegram.
However, this had no effect on the memory consumption.
The spell checker libraries were still seen in `/proc/$TELEGRAM_PID/map_files/`.

<details><summary>For example:</summary>
<pre>
/usr/lib/x86_64-linux-gnu/libaspell.so.15.2.0
/usr/lib/x86_64-linux-gnu/libhunspell-1.4.so.0.0.0
/usr/lib/x86_64-linux-gnu/libenchant.so.1.6.0
/usr/lib/x86_64-linux-gnu/enchant/libenchant_aspell.so
/usr/lib/x86_64-linux-gnu/enchant/libenchant_myspell.so
/usr/lib/x86_64-linux-gnu/enchant/libenchant_hspell.so
/usr/lib/x86_64-linux-gnu/enchant/libenchant_ispell.so
</pre>
</details>

See the full log in [problem.txt](https://github.com/telegramdesktop/tdesktop/files/4038118/problem.txt).

3.
Then I've uninstalled all the spell checkers and dictionaries I could find (using `apt-get purge` and `apt-get autoremove`), and restarted Telegram.
This solved the problem:  
* Telegram's physical memory consumption has decreased from ~2.5 GiB to ~100 MiB;  
* Telegram startup time has decreased to less than 1 second.

Also, I've noticed that the "Use system spell checker" knob has disappeared.

### Another way to reproduce

To verify that the problem was caused by the dictionaries and not by some random package I've just uninstalled, I did the following.

1. Re-installed `enchant` package (so that the "Use system spell checker" knob appeared again).
Note that the knob is still turned off.

2. Restarted Telegram and checked memory consumption (RES=~150 MiB).
See [without_dictionaries.txt](https://github.com/telegramdesktop/tdesktop/files/4038080/without_dictionaries.txt).

3. Installed 35 dictionary packages `myspell-$LOCALE`.
See [dictionaries_install_log.txt](https://github.com/telegramdesktop/tdesktop/files/4038082/dictionaries_install_log.txt).
See also [dictionary_packages.txt](https://github.com/telegramdesktop/tdesktop/files/4038086/dictionary_packages.txt) for the (supposedly) full list of installed spell-checking-related packages.

4. Restarted Telegram and checked memory consumption (RES=~1.2 GiB).
See [with_dictionaries.txt](https://github.com/telegramdesktop/tdesktop/files/4038089/with_dictionaries.txt).

5. Exited Telegram, and restarted it using the command `/home/oleg/bin/Telegram/Telegram` - just to have a look at Telegram's console output.
See [no_heaptrack_console.txt](https://github.com/telegramdesktop/tdesktop/files/4038090/no_heaptrack_console.txt).

6. Exited Telegram, and restarted it using the command `heaptrack /home/oleg/bin/Telegram/Telegram`.
See [heaptrack_console.txt](https://github.com/telegramdesktop/tdesktop/files/4038091/heaptrack_console.txt).

By the way, this time Telegram has segfaulted.
Despite this fact, I was able to analyze the log using `heaptrack_gui`.
The "Memory Peak" flame graph shows the call stack(s) of the allocation that have contributed to the peak memory consumption. As can be seen, the add_word() function, called via InitSpellChecker(), has allocated 590 MB (70%) of the peak memory consumption!

![heaptrack_gui](https://user-images.githubusercontent.com/59662050/72029177-a95bef80-32b7-11ea-8b75-5ccde98491b5.png)

**Comment by 23rd** [comment]
Thank you for the detailed report.

**Comment by farseerfc** [comment]
I think this is related to #6907 , because tdesktop tried to load all available dicts into memory.

**Comment by dm2912** [comment]
Same problem here in Mint 19.3

1.9.3 is using about 900mb

Cannot disable or remove any spell checkers as they are tied to cinnammon and if i try removing Mint threatens to remove itself :D

**Comment by joe247** [comment]
Happens in version v1.9.4... heavy CPU usage too just before GUI launch.
![Screenshot from 2020-01-18 19-34-54](https://user-images.githubusercontent.com/38415384/72665042-cec0ca00-3a2a-11ea-8324-ec52c350c021.png)

Also I don't know if I've to open a new issue (will be glad to do if so...) for this warning / error: 
```python
QApplication: invalid style override passed, ignoring it.
    Available styles: Windows, Fusion
Gtk-Message: 19:38:00.155: GtkDialog mapped without a transient parent. This is discouraged.
QTextCursor::setPosition: Position '13' out of range
QTextCursor::setPosition: Position '13' out of range
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x7fa8cc000900] Protocol name not provided, cannot determine if input is local or a network protocol, buffers and access patterns cannot be configured optimally without knowing the protocol
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x7fa90000ab00] Protocol name not provided, cannot determine if input is local or a network protocol, buffers and access patterns cannot be configured optimally without knowing the protocol
```

**Comment by 0136** [comment]
Quoting my original comment:

> 4. Restarted Telegram and checked memory consumption (RES=~1.2 GiB).
> See [with_dictionaries.txt](https://github.com/telegramdesktop/tdesktop/files/4038089/with_dictionaries.txt).

1.9.4 has fixed this for me. :)

Now the memory consumption (RES) just after Telegram startup does not exceed 135 000 KiB,
under the same conditions (with the same list of spelling/dictionary packages installed).

> I've turned off "System -> Advanced -> Use system spell checker", and restarted Telegram.
> However, this had no effect on the memory consumption.
> The spell checker libraries were still seen in `/proc/$TELEGRAM_PID/map_files/`.

1.9.4 has fixed this, too.

Now the spell checker libraries are seen in `map_files/` only if "Use system spell checker" is on  
(and only after I've typed something: looks like these libraries are loaded on demand).  
However, libenchant.so seems to be always loaded (even if "Use system spell checker" is off).

**1.9.4 memory usage on my laptop**

Step | Test                                                       | RES, KiB | Logs
---- | ---------------------------------------------------------- | -------- | ----
1    | Disabled "Use system spell checker" and restarted Telegram | ~121 000 | [log](https://github.com/telegramdesktop/tdesktop/files/4087509/spellcheck_off_started.txt)
2    | Typed a Russian word "Хорошо."                             | ~129 000 | [log](https://github.com/telegramdesktop/tdesktop/files/4087516/spellcheck_off_typed.txt)
3    | Enabled "Use system spell checker" and restarted Telegram  | ~121 000 | [log](https://github.com/telegramdesktop/tdesktop/files/4087518/spellcheck_on_started.txt)
4    | Typed a Russian word "Хорошо."                             | ~135 000 | [log](https://github.com/telegramdesktop/tdesktop/files/4087519/spellcheck_on_typed.txt)

After step 4, the word in the input box became marked with a red underline, which indicates that the spell checker works.
(Well, at least it tries to: I've spelled the word correctly, but it has marked it as incorrect.  
Anyway, I'm not a big fan of spell checkers, so I'll just disable it and be fine.)

***

@joe247, concerning your comment:

> Happens in version v1.9.4... heavy CPU usage too just before GUI launch.

Are you observing:  
A. high memory consumption and high CPU usage, or  
B. just high CPU usage?

On the screenshot you've attached I'm seeing B.

Regardless of the answer, I believe you should open a new issue against 1.9.4.

***

I'm closing this issue.  
Thanks to everyone. :)
